docs/requirements.md

```diff
@@ -0,0 +1,149 @@
+
+## Locale-sensitive layout
+
+Text layout is sensitive to locales, entailing a number of requirements. First is representation of locales. The underlying standard is [BCP 47]. We should try to use common Rust infrastructure for the parsing and representation. In addition, there are some utility functions that are likely specific to text, for example inferring script from other locale information (in ICU, this is done with the [`addLikelySubtags`] method).
+
+Locales manifest in two primary ways. First, they affect selection of fonts from a fallback list, especially Han unification. Second, they need to be plumbed to the shaping engine so OpenType [`locl`] features can be selected.
+
+Note that to really do this properly requires a locale *list,* not just a single locale as in some older interfaces. An example would be two users with en_US as their primary locale, but with ja_JP and zh_CN as their secondary locales. In this case, the secondary locale would be responsible for Han unification. In short, Han unification is controlled by the first CJK entry in the locale list. But Han unification is only the most visible example; in general, localization needs to be done from the first relevant entry, and the logic for what's relevant is complex and font-dependent.
+
+## Performance
+
+Text layout can be slow, and especially as more of the rendering pipeline is moving to the GPU, it can be the limiting factor. A large part of the design goal is performance, and here are some mechanisms to try to achieve that:
+
+* A layout cache. One subtlety is that it should be designed for multithreaded use, so a `Mutex<HashMap>` may replaced at some point with a concurrent hash map.
+
+* A fast representation of styles. Since style information can be expensive, both for storage and for hash-table lookup, we might design an arena, indexed by integer. Alternatively, we might use a small struct that covers the most common cases, with an optional `Box` for fancier style parameters such as OpenType features. In any case, locale is a large part of the puzzle.
+
+* Possibly fast paths for monospace fonts, including CJK. These have to be carefully designed to preserve semantics (unlike the "simple path" in Blink before it was removed).
+
+## Some open questions
+
+* Do shaping driven itemization, or use coverage data extracted from fonts (as in Minikin)?
+
+* Is layout at this level size-independent, or can it take into account pixel size dependent metrics (see [font-kit#34](https://github.com/pcwalton/font-kit/issues/34)).
+
+* Do we have a dependency on font-kit in all cases, or only cases where we're not using the platform layout? The font-kit dependency seems pretty heavyweight; more empirical measurement is necessary here.
```

Comment:
In my toys / experiments I’ve tried to see how far I could get with only Rust code and caller-provider fonts, ignoring system fonts completely. But that’s probably not realistic for most (?) programs, if only because shipping fonts with good Unicode coverage takes a lot of space.

That said, font-kit already has very different internals ("backends") based on the target platform. It’s not hard to imagine build-time configuration for alternative more lightweight implementation. (Although what would that look like?)

Reply:
What I'm getting at here is: do we use font-kit when the layout back-end is DirectWrite? A major consequence is whether we use font-kit types like `loaders::default::Font`, or abstract them. I'm leaning toward the latter, but open to discussion.

Reply:
Ah I see. What I was getting at is: font-kit is already an abstraction layer. Adding an abstraction layer that lets Skribo use font-kit or something else seems overkill. If we feel the need for it, maybe it’s a sign that font-kit is not at the right abstraction level in the first place?

Reply:
I'd like to centralize the platform font abstraction layers if we can. Font-kit has no users I'm aware of. If its design is problematic, I'm happy to scrap it and start over rather than to have yet another set of DirectWrite bindings to add to the stack.